lib/Frontend/ParseableInterfaceModuleLoader.cpp

```diff
@@ -0,0 +1,1040 @@
+  // directory above that.
+  return llvm::sys::path::parent_path(SpecificModuleCachePath);
+}
+
+#pragma mark - Forwarding Modules
+
+namespace {
+
+/// Describes a "forwarding module", that is, a .swiftmodule that's actually
+/// a YAML file inside, pointing to a different module but describing a
+/// different dependency resolution strategy.
+struct ForwardingModule {
+  /// The path to the original .swiftmodule in the prebuilt cache.
+  std::string underlyingModulePath;
+
+  /// Describes a set of file-based dependencies with their size and
+  /// modification time stored. This is slightly different from
+  /// \c SerializationOptions::FileDependency, because this type needs to be
+  /// serializable to and from YAML.
+  struct Dependency {
+    std::string path;
+    uint64_t size;
+    uint64_t lastModificationTime;
+  };
+  std::vector<Dependency> dependencies;
+  unsigned version = 1;
+
+  ForwardingModule() = default;
+  ForwardingModule(StringRef underlyingModulePath)
+  : underlyingModulePath(underlyingModulePath) {}
+
+  /// Adds a given dependency to the dependencies list.
+  void addDependency(StringRef path, uint64_t size, uint64_t modTime) {
+    dependencies.push_back({path.str(), size, modTime});
+  }
+};
+
+} // end anonymous namespace
+
+#pragma mark - YAML Serialization
+
+namespace llvm {
+  namespace yaml {
+    template <>
+    struct MappingTraits<ForwardingModule::Dependency> {
+      static void mapping(IO &io, ForwardingModule::Dependency &dep) {
+        io.mapRequired("mtime", dep.lastModificationTime);
+        io.mapRequired("path", dep.path);
+        io.mapRequired("size", dep.size);
+      }
+    };
+
+    template <>
+    struct SequenceElementTraits<ForwardingModule::Dependency> {
+      static const bool flow = false;
+    };
+
+    template <>
+    struct MappingTraits<ForwardingModule> {
+      static void mapping(IO &io, ForwardingModule &module) {
+        io.mapRequired("path", module.underlyingModulePath);
+        io.mapRequired("dependencies", module.dependencies);
+        io.mapRequired("version", module.version);
+      }
+    };
+  }
+} // end namespace llvm
+
+#pragma mark - Module Discovery
+
+namespace {
+
+/// The result of a search for a module either alongside an interface, in the
+/// module cache, or in the prebuilt module cache.
+class ModuleDiscovery {
+  /// The kind of module we've found.
+  enum class Kind {
+    /// A module that's either alongside the swiftinterface or in the
+    /// module cache.
+    Normal,
+
+    /// A module that resides in the prebuilt cache, and has hash-based
+    /// dependencies.
+    Prebuilt,
+
+    /// A 'forwarding' module.
+    /// \sa ForwardingModule.
+    Forwarding
+  };
+
+  /// The kind of module that's been discovered.
+  const Kind kind;
+
+  /// If the discovered module is a forwarding module, this stores the contents
+  /// of the forwarding file, once it's been loaded from disk.
+  const Optional<ForwardingModule> forwardingModule;
```

Comment:
Looking at how you're using ModuleDiscovery, I don't think you need to store this at all. You could just extract the prebuilt path at construction time.

(Also, should it be DiscoveredModule or something, so that it's a noun?)

Reply:
We also want to keep the separate dependencies, so we can validate them. I suppose we could do that inside this function, though. I want to keep the dependency validation logic together, though.

Reply:
I thought we validated the dependencies before we actually built the ModuleDiscovery, but maybe I'm wrong.

Reply:
Actually, you're right. DiscoveredModule should not hold the forwarding module -- we can separate loading/checking dependencies of a forwarding module from the results of module discovery.